presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaPageSink.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.kafka;
+
+import com.google.common.collect.ImmutableList;
+import io.airlift.slice.Slice;
+import io.prestosql.plugin.kafka.encoder.EncoderFieldValueProvider;
+import io.prestosql.plugin.kafka.encoder.RowEncoder;
+import io.prestosql.spi.Page;
+import io.prestosql.spi.connector.ConnectorPageSink;
+import io.prestosql.spi.type.Type;
+import org.apache.kafka.clients.producer.KafkaProducer;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.CompletableFuture.completedFuture;
+
+public class KafkaPageSink
+        implements ConnectorPageSink
+{
+    private final String topicName;
+    private final List<KafkaColumnHandle> columns;
+    private final List<Type> columnTypes;
+    private final KafkaProducerFactory producerFactory;
+    private final RowEncoder keyEncoder;
+    private final RowEncoder messageEncoder;
+
+    public KafkaPageSink(
+            KafkaProducerFactory producerFactory,
+            String topicName,
+            List<KafkaColumnHandle> columns,
+            RowEncoder keyEncoder,
+            RowEncoder messageEncoder)
+    {
+        this.producerFactory = requireNonNull(producerFactory, "producerFactory is null");
+        this.topicName = requireNonNull(topicName, "topicName is null");
+        this.columns = ImmutableList.copyOf(requireNonNull(columns, "columnNames is null"));
+        this.keyEncoder = requireNonNull(keyEncoder, "keyEncoder is null");
+        this.messageEncoder = requireNonNull(messageEncoder, "messageEncoder is null");
+
+        this.columnTypes = columns.stream()
+                .map(KafkaColumnHandle.class::cast)
+                .map(KafkaColumnHandle::getType)
+                .collect(toImmutableList());
+    }
+
+    @Override
+    public CompletableFuture<?> appendPage(Page page)
```

Comment:
That is mean we publish each page to kafka directly, as a set of rows?

- Is there any way to 'stage' produces messages and make them visible upon `finish` (or `ConnectorMetadata#finishInsert`)?
  - add a code comment explaining design decision
- should there be any buffering?

Reply:
The producer does some buffering internally (configurable).  To make changes visible on `finishInsert()`, we'd have to start a transaction on `beginInsert()`, make sure that all producers share the same `transaction.id` (probably the Presto query id) and commit in `finishInsert()`.

I'm not a Kafka expert, so we felt we could use more time to think about this (feedback welcome) and some other issues.  That was intended for a follow-up PR (this one was supposed be sort of an MVP; it came out larger than I expected).

Reply:
I'll look into this, also have to make it possible to send messages to different partitions

Reply:
> The producer does some buffering internally (configurable).

currently it's short-lived

Reply:
Producer lives in worker while `finishInsert()` is called on coordinator.